File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    public static final String BACKSTAGE_PASSES_TO_A_TAFKAL_80_ETC_CONCERT = "Backstage passes to a TAFKAL80ETC concert";
    public static final String SULFURAS_HAND_OF_RAGNAROS = "Sulfuras, Hand of Ragnaros";
    public static final String AGED_BRIE = "Aged Brie";

    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    private static void updateItem(Item item) {
        updateQuality(item);
        updateExpiration(item);
        if (isExpired(item)) {
            processExpired(item);
        }
    }

    private static void updateQuality(Item item) {
        if (item.name.equals(AGED_BRIE)) {
            increaseQuality(item);
        } else if (item.name.equals(BACKSTAGE_PASSES_TO_A_TAFKAL_80_ETC_CONCERT)) {
            increaseQuality(item);
            if (item.sellIn < 11) {
                increaseQuality(item);
            }

            if (item.sellIn < 6) {
                increaseQuality(item);
            }
        } else if (!item.name.equals(SULFURAS_HAND_OF_RAGNAROS)) {
            decreaseQuality(item);
        }
    }

    private static void updateExpiration(Item item) {
        if (item.name.equals(SULFURAS_HAND_OF_RAGNAROS)) {
            return;
        }
        item.sellIn = item.sellIn - 1;
    }

    private static boolean isExpired(Item item) {
        return item.sellIn < 0;
    }

    private static void processExpired(Item item) {
        if (item.name.equals(AGED_BRIE)) {
            increaseQuality(item);
        } else if (item.name.equals(BACKSTAGE_PASSES_TO_A_TAFKAL_80_ETC_CONCERT)) {
            item.quality = 0;
        } else if (!item.name.equals(SULFURAS_HAND_OF_RAGNAROS)) {
            decreaseQuality(item);
        }
    }

    private static void decreaseQuality(Item item) {
        if (item.quality > 0) {
            item.quality = item.quality - 1;
        }
    }

    private static void increaseQuality(Item item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }
    }

    public void updateInventory() {
        for (Item item : items) {
            updateItem(item);
        }
    }
}

File: go/gildedrose/gildedrose.go
package gildedrose

type Item struct {
	Name            string
	SellIn, Quality int
}

const (
	AgedBrie      = "Aged Brie"
	BackStagePass = "Backstage passes to a TAFKAL80ETC concert"
	Sulfuras      = "Sulfuras, Hand of Ragnaros"
)

func ProcessDailyItems(items []*Item) {
	for _, item := range items {
		updateItemQuality(item)
		updateExperation(item)
		if item.SellIn < 0 {
			processExpiredItem(item)
		}
	}

}

func updateItemQuality(item *Item) {

	if item.Name == "Aged Brie" {
		increaseQuality(item)
	} else if item.Name == BackStagePass {
		increaseQuality(item)
		if item.SellIn < 11 {
			increaseQuality(item)
		}
		if item.SellIn < 6 {
			increaseQuality(item)
		}

	} else if item.Name == Sulfuras {
		return
	} else {
		decreaseQuality(item)
	}
}

func decreaseQuality(item *Item) {
	if item.Quality > 0 {
		item.Quality = item.Quality - 1
	}
}

func updateExperation(item *Item) {
	if item.Name == Sulfuras {
		return
	} else {
		item.SellIn = item.SellIn - 1
	}
}

func processExpiredItem(item *Item) {
	if item.Name == AgedBrie {
		increaseQuality(item)
	} else if item.Name == BackStagePass {
		item.Quality = 0
	} else if item.Name == Sulfuras {
		return
	} else {
		decreaseQuality(item)
	}
}

func increaseQuality(item *Item) {
	if item.Quality < 50 {
		item.Quality = item.Quality + 1
	}
}
